File: lib/snake_engine.rb
#!/usr/bin/env ruby
require 'sdl'
require 'abstract_state'
require 'gameover_state'
require 'highscore_state'
require 'multi_state'
require 'single_state'

module JakeTheSnake
  class Snake_Engine
    attr_accessor :tick_interval, screen

    def initalize(width, height, fullscreen)
      done = false
      $image_buffer = Hash.new
      state = Abstract_State.new
      if(SDL::Init(SDL::INIT_VIDEO) == -1)
        puts "Can't init SDL: #{SDL::SDL_GetError}"
        exit 1
      end

      if fullscreen
        flags = SDL::SDL_SWSURFACE or SDL::SDL_FULLSCREEN
      else
        flags = SDL::SDL_SWSURFACE
      end

      if screen == nil
        puts "Can't set video mode: #{SDL::SDL_GetError}"
        SDL::SDL_Quit
        exit 1
        state = nil
        tick_interval = 50
      end

      def set_state(state)
        self.state = nil
        self.state = state
        self.state.set_engine(self)
      end

      def loop
        event = SDL::SDL_event

        if state == nil
          return
        end

        while not done
          next_tick = SDL::SDL_GetTicks.to_i + tick_interval

          while SDL::SDL_PollEvent(event)
            case event.type
            when SDL::SDL_Quit
              quit
              break
            when SDL::SDL_KEYDOWN
              state.key_pressed(event.key.keysym.sym)
              break
            end
          end
          state.clock_tick
          state.draw(screen)

          SDL::SDL_Flip(screen)

          if SDL::SDL_GetTicks < next_tick
            SDL::SDL_Delay(next_tick - SDL_GetTicks)
          end
          SDL::SDL_Quit
        end

        def quit
          done = true
        end

        def load_image(filename)
          if image_buffer.find(filename) != image_buffer.end
            # If the image already has been loaded, just return.
            exit 1
          end

          loaded_image = SDL::Surface = 0
          optimized_image = SDL::Surface = 0
          loaded_image = SDL::SDL_LoadBMP(filename)

          if loaded_image
            optimized_image = SDL_DisplayFormat(loaded_image)
            SDL::SDL_FreeSurface(loaded_image)
          end
          optimized_image
        end
      end
    end
  end
end

File: c++_src/Snake_Engine.cpp
#include "Snake_Engine.h"

bool Snake_Engine::done = false;
std::map<std::string, SDL_Surface*> Snake_Engine::image_buffer;

/* ------------- Class Snake_Engine ------------- */

Snake_Engine::Snake_Engine(int width, int height, bool fullscreen) {
	Uint32 flags;
	if (-1 == SDL_Init(SDL_INIT_VIDEO)) {
		printf("Can't init SDL:  %s\n", SDL_GetError());
		exit(1);
	}
	if (fullscreen) {
		flags = SDL_SWSURFACE | SDL_FULLSCREEN;

	} else {
		flags = SDL_SWSURFACE;
	}
	screen = SDL_SetVideoMode(width, height, 32, flags);
	if (NULL == screen) {
		printf("Can't set video mode: %s\n", SDL_GetError());
		SDL_Quit();
		exit(1);
	}
	state = NULL;
	tick_interval = 50; // 20 FPS
}

Snake_Engine::~Snake_Engine() {
	// No need to delete state as this object is always active
	// Deleting object casues segmentation fault due to this.
	// We dont have any deallocation here because this object is
	// always alive and memory is taken back by other classes.
}

void Snake_Engine::set_state(Abstract_State* state) {
	delete this->state;
	this->state = 0;
	this->state = state;
	this->state->set_engine(this);
}

void Snake_Engine::set_tick_interval(int interval) {
	tick_interval = interval;
}

SDL_Surface* Snake_Engine::get_screen() const {
	return screen;
}

void Snake_Engine::loop() {
	SDL_Event event;
	Uint32 next_tick;

	if (NULL == state) {
		return;
	}
	while (!done) {
		next_tick = SDL_GetTicks() + tick_interval;

		while (SDL_PollEvent(&event)) {
			switch (event.type) {
			case SDL_QUIT:
				quit();
				break;
			case SDL_KEYDOWN:
				state->key_pressed(event.key.keysym.sym);
				break;
			}
		}
		state->clock_tick();
		state->draw(screen);

		SDL_Flip(screen);

		if (SDL_GetTicks() < next_tick) {
			SDL_Delay(next_tick - SDL_GetTicks());
		}
	}
	SDL_Quit();
}

void Snake_Engine::quit() {
	done = true;
}

SDL_Surface* Snake_Engine::load_image(const std::string& filename)  {
	if(image_buffer.find(filename) != image_buffer.end())
		// If the image already has been loaded, just return.
		return image_buffer[filename];
	SDL_Surface* loaded_image = 0;
	SDL_Surface* optimized_image = 0;
	loaded_image = SDL_LoadBMP(filename.c_str());
	if (loaded_image != 0) {
		optimized_image = SDL_DisplayFormat(loaded_image);
		SDL_FreeSurface(loaded_image);
	}
	image_buffer.insert(make_pair(filename, optimized_image));
	return optimized_image;
}

int Snake_Engine::rand_x() {
	// Randoms out a x coordinate within the green background.
	return (rand() % 38 * 16 + 16);
}

int Snake_Engine::rand_y() {
	// Randoms out a y coordinate within the green background.
	return (rand() % 28 * 16 + 16);
}

void Snake_Engine::free_memory(SDL_Surface* surface) {
	if (surface !=0) {
		SDL_FreeSurface(surface);
		surface = 0;
	}
}
